### ISSUE TITLE
`scipy.sparse.sparsetools` deprecated

### ISSUE THREAD

**Issue by piskvorky** [RaRe-Technologies/gensim#462]
Importing gensim prints this deprecation warning:

``` bash
gensim/home/ubuntu/.vew/ds26/lib/python2.6/site-packages/numpy/lib/utils.py:95: DeprecationWarning: `scipy.sparse.sparsetools` is deprecated!
scipy.sparse.sparsetools is a private module for scipy.sparse, and should not be used.
  warnings.warn(depdoc, DeprecationWarning)
```

Find out why it's deprecated and, most importantly, what is the replacement?

We rely on sparsetools rather heavily in gensim, so this is critical.

Resources: scipy release notes, scipy mailing list...

**Comment by tmylk** [comment]
Pauli Virtanen from scipy mailing list has suggested the following:

>If I understand correctly, these are use cases that can be expressed
>in terms of usual sparse matrix operations,

>y = corpus * o
>y += corpus * chunk

>but you are using the internal sparsetools routines instead,
>because of performance reasons? Is the performance difference
>big in this case? Is the issue that you want in-place sparse AXPY,
>or is it due to dealing with small matrices and avoiding
>overheads?

>There's currently no sparse axpy available in Scipy.
>There probably should be though.

>Gensim is not a pure-Python module, so one relatively straightforward
>possibility is to just bundle a copy of the current sparsetools
>module (or just the one routine you need) with it. There's no SWIG
>nowadays involved, and it's independent of the rest of Scipy,
>so it's probably doable.



https://mail.scipy.org/pipermail/scipy-user/2015-October/036712.html

**Comment by piskvorky** [comment]
Did you check that `corpus * o` doesn't materialize the sparse matrix `corpus` into a dense array?

If it's memory efficient (no extra memory overhead) then we can change it. I don't care (much) about performance here, I'm pretty sure this optimization was concerned with memory, not time.  We cannot afford to convert `corpus` into a dense array.

Gensim is pure-python, extensions are only optional, so bundling a hard dependency that needs compilation is not an option for us.

**Comment by piskvorky** [comment]
And it looks like `y += corpus * chunk` would double the memory in any case (a temporary matrix for `corpus * chunk` before adding it to the existing `y`) :(

**Comment by tmylk** [comment]
Feature request submitted to sci-py for sparse in-place AXPY

https://github.com/scipy/scipy/issues/5348

**Comment by tmylk** [comment]
There is some progress happening on scipy

**Comment by souravsingh** [comment]
The issue is currently in works in this repo here- https://github.com/souravsingh/sparseutils
Finishing this should also fix #557

**Comment by menshikh-iv** [comment]
Can't reproduce now, continue the discussion in #557.
